### DISCUSSION TITLE
Running cypress GUI in <iframe>

### DISCUSSION THREAD

**Discussion by OliverLjung** [cypress-io/cypress#28390] [Questions and Help]
Hi!

This is my first time trying Cypress and I have been experimenting for the last 10 hours or so; very cool stuff 👨‍💻 
I had an idea that I would create a service that runs my cypress tests. My idea is to:

1. Build a webapp (I use python flask for now)
2. Create and expose an endpoint that takes a cypress script in plain text as input (http)
3. From the app, create the `*.cy.js` file with the input cypress script as content
4. Run the cypress code. I do this with pythons `asyncio.create_subprocess_shell` like this 
```python
async def run_cypress(port, spec):
    await asyncio.create_subprocess_shell(
        cmd=f'npx cypress run --headed --no-exit --port {port} --browser chrome --spec {spec}'
    )
```
5. return the template which contains my `<iframe>`

So far, steps 1-4 work and I can see my test example (spec) running as it should.
The problem I want to solve is how to access Cypress GUI in the `<iframe>`. 

I have tried playing around with the different CLI options and tried to access it with `localhost:port` which has only come this far..
![image](https://github.com/cypress-io/cypress/assets/64260361/a1cf5391-e596-49aa-9fa9-b0a91d9cfe47)

Am I trying to do something that shouldn't be possible? 
Any help and/or pointer are greatly appreciated 😄

**Comment by portyu9** [comment]
Hi @OliverLjung 

`--port` does **not** turn the Cypress Runner into a web application that you can embed with:

```html
<iframe src="http://localhost:1234"></iframe>
```

That port is part of Cypress's internal test/proxy infrastructure. It is not a supported public URL for hosting the Runner UI.

So this:

```bash
npx cypress run --headed --no-exit --port 1234
```

can show Cypress on the machine where Cypress is running, but it does not expose that GUI as an embeddable HTTP service.

If your goal is a remote Cypress-running service, I would separate:

1. **Execution**
   - accept/select the spec;
   - launch Cypress on the server;
   - capture stdout/results/screenshots/video.

2. **Presentation**
   - render those results in your own Flask UI.

If you need users to interact with the actual remote Cypress desktop/browser, use a remote-display technology such as VNC/noVNC or a browser-accessible desktop around the machine/container running Cypress. Current Cypress documentation uses that pattern for browser-accessible container desktops.

That is different from embedding Cypress's Runner DOM directly in your application's `<iframe>`.

One additional caution: accepting arbitrary Cypress/JavaScript source over HTTP and executing it on your server is effectively arbitrary-code execution. If this service accepts untrusted users, run every job in a tightly isolated disposable container/VM with restricted credentials, filesystem and network access.

Please mark this answer as accepted if it helped, thank you.
